nemo/collections/speechlm/recipes/pipeline.py

```diff
@@ -0,0 +1,170 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from omegaconf import DictConfig, OmegaConf
+
+from nemo import lightning as nl
+from nemo.collections.common.tokenizers import AutoTokenizer
+from nemo.collections.speechlm.api import _setup, finetune
+from nemo.collections.speechlm.data.audio_to_text_module import AudioToTextDataModule
+from nemo.collections.speechlm.models.speech_to_text_llm_model import SpeechToTextLLM, SpeechToTextLLMConfig
+from nemo.collections.speechlm.modules.asr_module import ASRModuleConfig
+from nemo.collections.speechlm.modules.modality_adapter import ModalityAdapterConfig
+from nemo.collections.speechlm.utils import SpeechToTextLLMPEFT, get_object_list_from_config
+from nemo.core.classes.common import Serialization, typecheck
+from nemo.lightning.pytorch.callbacks import PreemptionCallback
+from nemo.utils import logging
+from nemo.utils.exp_manager import StatelessTimer, TimingCallback
+
+
+def speech_to_text_llm_train(cfg: DictConfig):
+    """Train the model using provided config."""
+    typecheck.set_typecheck_enabled(enabled=False)  # disable typechecks from NeMo 1.x
+    cfg = OmegaConf.to_container(cfg, resolve=True)
+
+    logging.info(f'Hydra config: {OmegaConf.to_yaml(cfg)}')
+
+    # 1. build the model
+    tokenizer = AutoTokenizer(cfg['model']['llm']['pretrained_model'])
+    model_config = SpeechToTextLLMConfig(
+        language_model_class=cfg['model']['llm']['_target_'],
+        language_model_config=Serialization.from_config_dict(cfg['model']['llm']['config']),
+        speech_model_config=ASRModuleConfig(**cfg['model']['speech_encoder']),
+        modality_adapter_config=ModalityAdapterConfig(**cfg['model']['modality_adapter']),
+        language_model_from_pretrained=cfg['model']['llm']['pretrained_model'],
+        freeze_language_model=cfg['model']['freeze_language_model'],
+        freeze_speech_model=cfg['model']['freeze_speech_model'],
+        freeze_modality_adapter=cfg['model']['freeze_modality_adapter'],
+        data_config=cfg['data']['common'],
+    )
+
+    model = SpeechToTextLLM(config=model_config, tokenizer=tokenizer)
+
+    # 2. build dataset
+    data = AudioToTextDataModule(cfg['data'], tokenizer=tokenizer)
+
+    # 3. setup the optimizer
+    optim = Serialization.from_config_dict(cfg['optim'])
+
+    # 4. setup trainer
+    callbacks = get_object_list_from_config(cfg['callbacks'])
+    if cfg.get('max_time_per_run', None):
+        if cfg['strategy'].get('ckpt_async_save', True):
+            logging.warning(
+                f"`strategy.ckpt_async_save` must be `False` to save ckpt when `max_time_per_run` is set,",
+                f"got {cfg['strategy']['ckpt_async_save']}. `max_time_per_run` will not work in this case!",
```

Comment:
Better to error out here. I dont want a run to go on assuming it will stop at max time, only to get a corrupted checkpoint after 4 hours

Reply:
updated to raise error